src/dune_rules/fetch_rules.ml

```diff
@@ -0,0 +1,275 @@
+;;
+
+let gen_rules ctx ~dir ~components =
+  match components with
+  | [] ->
+    Memo.return Rules.empty
+    |> Gen_rules.make
+         ~build_dir_only_sub_dirs:
+           (Gen_rules.Build_only_sub_dirs.singleton ~dir Subdir_set.all)
+    |> Memo.return
+  | [ checksum ] ->
+    let checksum = Dune_pkg.Checksum.parse_string_exn (Loc.none, checksum) in
+    let directory_targets =
+      let target_dir = make_target ~kind:`Directory checksum in
+      Path.Build.Map.singleton target_dir Loc.none
+    in
+    let rules =
+      Rules.collect_unit (fun () ->
+        url_of_checksum ctx checksum
+        >>= function
+        | None ->
+          User_error.raise
+            [ Pp.textf "unknown checksum %s" (Checksum.to_string checksum) ]
+        | Some (url, checksum) -> gen_rules_checksum url checksum)
+    in
+    Gen_rules.make rules ~directory_targets |> Memo.return
+  | _ -> Memo.return Gen_rules.no_rules
+;;
+
+module Copy = struct
+  module Spec = struct
+    type ('path, 'target) t =
+      { src_dir : 'path
+      ; dst_dir : 'target
+      }
+
+    let name = "copy-dir"
+    let version = 1
+    let bimap t f g = { src_dir = f t.src_dir; dst_dir = g t.dst_dir }
+    let is_useful_to ~memoize = memoize
+
+    let encode { src_dir; dst_dir } path target =
+      Dune_lang.List
+        [ Dune_lang.atom_or_quoted_string name; path src_dir; target dst_dir ]
+    ;;
+
+    let action { src_dir; dst_dir } ~ectx:_ ~eenv:_ =
+      let open Fiber.O in
+      let+ () = Fiber.return () in
+      Path.mkdir_p (Path.build dst_dir);
+      let dst_dir = Path.build dst_dir in
+      Fpath.traverse
+        ~init:()
+        ~dir:(Path.to_string src_dir)
+        ~on_dir:(fun ~dir fname () ->
+          Path.L.relative dst_dir [ dir; fname ] |> Path.mkdir_p)
+        ~on_file:(fun ~dir fname () ->
+          let src = Path.L.relative src_dir [ dir; fname ] in
+          let dst = Path.L.relative dst_dir [ dir; fname ] in
+          Io.copy_file ~src ~dst ())
+    ;;
+  end
+
+  let action ~src_dir ~dst_dir =
+    let module M = struct
+      type path = Path.t
+      type target = Path.Build.t
+
+      module Spec = Spec
+
+      let v = { Spec.dst_dir; src_dir }
+    end
+    in
+    Action.Extension (module M)
+  ;;
+end
+
+let deps kind checksum =
+  let src = make_target ~kind checksum in
+  Path.build src |> Dep.file |> Action_builder.dep |> Action_builder.with_no_targets
+;;
+
+let copy ~target kind checksum =
+  let src = Path.build (make_target ~kind checksum) in
+  let open Action_builder.With_targets.O in
+  deps kind checksum
+  >>>
+  match kind with
+  | `File -> Action_builder.copy ~src ~dst:target
+  | `Directory ->
+    Copy.action ~src_dir:src ~dst_dir:target
+    |> Action.Full.make
+    |> Action_builder.With_targets.return
+    |> Action_builder.With_targets.add_directories ~directory_targets:[ target ]
+;;
+
+let fetch ~target kind url checksum =
+  match checksum with
+  | Some (_, checksum) -> copy ~target kind checksum
```

Comment:
Will this not fail when a checksum is specified but not present in the cache?

Reply:
It will not. It will introduce a dependency on `_fetch/$checksum/<file|dir>` which will load the rules in that directory and generate the rule for this checksum (if it's a valid checksum).